### DISCUSSION TITLE
How to improve unocss performance in the browser

### DISCUSSION THREAD

**Discussion by oeyoews** [unocss/unocss#3117] [Q&A]
The unocss runtime seems to query all elements. If there are many elements in a single html (such as this https://tiddlywiki.com/), the page will get stuck; but if using the CDN of tailwindcss does not have this problem, can we improve this?

ps: There doesn’t seem to be much difference between using tailwindcss or unocss in a single file.

![image](https://github.com/unocss/unocss/assets/72405338/d2521fae-d14d-437e-9d41-30e172f77fc7)

**Comment by antfu** [comment]
We could refactor the code to use loops instead of recursions. PR welcome :)

**Comment by Schmavery** [comment]
Thanks for posting this! I was at loss for why my tiddlywiki was hanging/crashing while trying to add the unocss runtime.
I'm sure you've moved on by now, but in case anyone else runs into this, I had some success setting the root element to ignore the tiddler store areas

```html
<script>
window.__unocss = {
    // ...
    runtime: {
        rootElement: () => document.getElementsByClassName("tc-page-container-wrapper")[0],
    }
}
</script>
```

**Reply by oeyoews** [reply]
cool, I will try it
